dirty_cat/super_vectorizer.py

```diff
@@ -0,0 +1,341 @@
+                """
+                Method to call on each column,
+                which will try to cast said column to the best possible type.
+                """
+
+                # Check if the array contains NaN or INF
+                if _has_missing_values(col):
+                    raise ValueError(_ERR_MSG_FOUND_NAN)
+                if any(col.isinf()):
+                    raise ValueError(_ERR_MSG_FOUND_INF)
+
+                return cast_astype(col)
+
+            try:
+                # nD array
+                arr = np.apply_along_axis(func1d=cast_column, axis=1, arr=arr)
+            except np.AxisError:
+                # 1D array
+                arr = cast_column(arr)
+
+            return arr
+
+        if isinstance(X, pd.Series):
+            return cast_series(X)
+        elif isinstance(X, pd.DataFrame):
+            return cast_dataframe(X)
+        elif isinstance(X, np.ndarray):
+            return cast_array(X)
+
+    def _transform(self, X) -> pd.DataFrame:
+        # Convert to pandas DataFrame if not already.
+        if not isinstance(X, pd.DataFrame):
+            X = pd.DataFrame(X)
+
+        # Detect if the array contains missing values.
+        if _has_missing_values(X):
+            if self.handle_missing == '':
+                X = _replace_missing(X)
+            elif self.handle_missing == 'error':
+                raise ValueError('Array contains missing values')
+            else:
+                raise ValueError("Invalid 'handle_missing' value. "
+                                 "Expected any of {'', 'error'}, "
+                                 f"got {self.handle_missing}")
+
+        if self.auto_cast:
+            X = self._auto_cast_array(X)
+
+        return X
+
+    def transform(self, X):
+        X = self._transform(X)
+        # [Black magic] - Calls the overridden method `transform`.
+        return ColumnTransformer.transform(self, X)
+
+    def fit_transform(self, X, y=None):
+        X = self._transform(X)
+
+        # Select columns by dtype
+        numeric_columns = X.select_dtypes(include=['int', 'float']).columns.to_list()
+        string_columns = X.select_dtypes(include=['string', 'object']).columns.to_list()
+        categorical_columns = X.select_dtypes(include='category').columns.to_list()
+        datetime_columns = X.select_dtypes(include='datetime').columns.to_list()
+
+        # Divide string and categorical columns by cardinality
+        low_card_str_columns = [col for col in string_columns if X[col].nunique() < self.cardinality_threshold]
+        high_card_str_columns = [col for col in string_columns if X[col].nunique() >= self.cardinality_threshold]
+        low_card_cat_columns = [col for col in categorical_columns if X[col].nunique() < self.cardinality_threshold]
+        high_card_cat_columns = [col for col in categorical_columns if X[col].nunique() >= self.cardinality_threshold]
+
+        # Construct the transformers
```

Comment:
I don't really understand what the code below is doing (forgive my tired brain).

These lines need at least a high-level comment before them describing them.

Also, I am surprised, it seems that they are using a dataframe to build a list. I wonder whether it would not be more readable using a combination of for loops and the append or extend method of lists.

Reply:
Sure, I'll try to improve the doc  
I used this process because it was the lightest code-wise. I'll implement the same thing in vanilla Python and compare the two.